File: patterns-java/src/main/java/me/freedom4live/patterns/java/factory/impl/CatFactoryImpl.java
package me.freedom4live.patterns.java.factory.impl;

import me.freedom4live.patterns.java.factory.Cat;
import me.freedom4live.patterns.java.factory.CatFactory;

import java.util.UUID;

public class CatFactoryImpl implements CatFactory {
    @Override
    public Cat createRandomCat() {
        return new Cat() {

            private String name = UUID.randomUUID().toString();

            @Override
            public String meow() {
                return name;
            }
        };
    }

    @Override
    public Cat createNamedCat(String name) {
        return () -> name;
    }
}

File: patterns-kotlin/src/main/kotlin/me/freedom4live/patterns/kotlin/factory/impl/CatFactoryImpl.kt
package me.freedom4live.patterns.kotlin.factory.impl

import me.freedom4live.patterns.kotlin.factory.Cat
import me.freedom4live.patterns.kotlin.factory.CatFactory
import java.util.*

class CatFactoryImpl : CatFactory {
    override fun randomCat(): Cat = object : Cat {
        private val name = UUID.randomUUID().toString()
        override fun meow(): String = name
    }

    override fun namedCat(name: String): Cat = object : Cat {
        override fun meow(): String = name
    }
}
